docs/proposals/20230612-node-os-image-caching.md

```diff
@@ -0,0 +1,260 @@
+
+We will know we’ve succeeded when we can benchmark speed increases and successful image changes.
+
+### Non-Goals/Future Work
+
+1. Extend the functionality to Windows nodes
+1. Optimization for efficiency and scalability
+1. A more complicated method of selecting a good candidate node such as incorporating manual forced prototype creation, which would perform better but take research to find the optimal method
+1. Optimization of the default time interval into a specific, best general default which would have to be researched
+1. Automatic bad snapshot rollbacks since it will be hard to generically determine when a snapshot is bad currently
+1. Annotating the timestamp on each node of when it was updated with automatic security updates
+1. Allow the feature to be enabled for any type of image (marketplace or custom with ID)
+1. AzureMachineTemplate support, holding out for now because of immutability of the AzureMachineTemplates
+
+## Proposal
+
+### User Stories
+
+#### As an operator, I want to be able to have faster horizontal scaling with pods running quicker when scheduled
+
+As a cluster operator I want to bring up new nodes that already have the container images of the applications it will run pre-cached so that pods start running quicker when scheduled, resulting in faster horizontal scaling.
+
+#### As an operator of clusters managed by CAPZ, I want to reduce monetary and environmental costs
+
+As a cluster operator I want to have lower monetary and environmental costs. I want my Nodes' OS images to be continually cached, and thus I will be able to avoid longer pull times and avoid having to create warm nodes and overprovision. This will be better for the environment and cheaper for me as I will not have to waste excess resources on warm nodes.
+
+#### As an operator, I want to be able to be more inline with security compliance as I spin up nodes
+
+As an operator I would like to be able to have my node’s OS image cache for the ability to avoid security alerts, cache security updates and patches so that I’ll be more security compliant and up to date when I have to spin up a new node. Otherwise, flags will be raised that a node is out of date on boot up before it finishes patching and a temporary security risk will exist.
+
+### Implementation Details/Notes/Constraints
+
+The plan is to modify the existing Controllers with the Node Prototype Pattern as desired. These controller additions can be added to AzureMachinePool Controller.
+
+An operator will be able to decide to turn the feature on or off with an environment variable on clusterctl initialization, and then on a cluster by cluster basis can alter the AzureMachinePools to specify the use of the feature. They can also update the AzureMachinePool to customize how long they want the caching interval to be (see the yaml files below in this section for the caching interval).
+
+Example of the environment variable being turned on:
+
+```
+export AZURE_OS_CACHING=true
```

Comment:
```suggestion
export AZURE_NODE_OS_CACHING=true
```

this is arguably less ambiguous (although longer)

Reply:
Sounds good to me